codice/ddf-ui
# [4.1.x] Adds USNG/MGRS, DMS, and UTM/UPS input to Line/Poly for Search

_Requires usng.js updates: https://github.com/codice/usng.js/commit/33eab1c4d383996018a9486fc6c383abd66274a8_

**What does this PR do?**
Provides DMS, USNG/MGRS, and UTM/UPS input for line and polygon options in the search pane. 
New Features:
- Degrees, Minutes, Seconds Input for line/polygon location searches
- USNG/MGRS Input for line/polygon location searches
  - Falls back to UPS for areas around poles
- UTM/UPS Input for line/polygon location searches

**How to Test**
(Use 2D map and 3D map for all of these)
- Check the Location checkbox on the search page
- Select "Line" from the dropdown
- With DD selected, click draw and draw a line on the map
- Click through the other types and make sure the line was properly translated
- Click through the other types and try typing in the points, make sure validation works, make sure typing live updates the map, make sure that you can add and delete points
- Navigate away from the search page and renavigate back and select location and line again
- For each of the types (DMS, USNG, UTM/UPS), make sure you can click the + button to add new points and type in and after you have a valid shape it appears on the map

- Repeat these steps for polygon!

![Screen Shot 2021-02-24 at 14 11 54](https://user-images.githubusercontent.com/9013407/109081989-624ddb00-76c0-11eb-9d45-5ec23ce3909c.png)
![Screen Shot 2021-02-24 at 14 12 02](https://user-images.githubusercontent.com/9013407/109081994-637f0800-76c0-11eb-8d1d-d90b6e49d6d9.png)
![Screen Shot 2021-02-24 at 14 12 36](https://user-images.githubusercontent.com/9013407/109081996-64b03500-76c0-11eb-8992-bdb08bd1766e.png)
![Screen Shot 2021-02-24 at 14 12 47](https://user-images.githubusercontent.com/9013407/109082001-6548cb80-76c0-11eb-8c4e-088eafc16547.png)
